quinn/src/connection.rs

```diff
@@ -79,29 +79,47 @@ impl Connecting {
+    /// cryptographic protection.
     ///
-    /// When the `ZeroRttAccepted` future completes, the connection has been fully established.
+    /// ## Outgoing
     ///
-    /// # Security
+    /// For outgoing connections, the initial attempt to convert to 0-RTT will proceed if the
+    /// [`crypto::ClientConfig`][crate::crypto::ClientConfig] attempts to resume a previous TLS
+    /// session. However, **the remote endpoint may not actually _accept_ the 0-RTT data**--yet
+    /// still accept the connection attempt in general. This possibility is conveyed through the
+    /// [`ZeroRttAccepted`] future--when the handshake completes, it resolves to true if the 0-RTT
+    /// data was accepted and false if it was rejected. If it was rejected, streams opened prior
+    /// to the handshake completing will _hang_, and their existence will typically not be conveyed
+    /// to the remote application.
     ///
-    /// On outgoing connections, this enables transmission of 0-RTT data, which might be vulnerable
-    /// to replay attacks, and should therefore never invoke non-idempotent operations.
+    /// A server may reject 0-RTT data at its discretion, but accepting 0-RTT data requires the
+    /// relevant resumption state to be stored in the server, which servers may limit or lose for
+    /// various reasons including not persisting resumption state across server restarts.
     ///
-    /// On incoming connections, this enables transmission of 0.5-RTT data, which might be
-    /// intercepted by a man-in-the-middle. If this occurs, the handshake will not complete
-    /// successfully.
+    /// If manually providing a [`crypto::ClientConfig`][crate::crypto::ClientConfig], check your
+    /// implementation's docs for 0-RTT pitfalls.
     ///
-    /// # Errors
+    /// ## Incoming
     ///
-    /// Outgoing connections are only 0-RTT-capable when a cryptographic session ticket cached from
-    /// a previous connection to the same server is available, and includes a 0-RTT key. If no such
-    /// ticket is found, `self` is returned unmodified.
+    /// For incoming connections, conversion to 0.5-RTT will always fully succeed. `into_0rtt` will
+    /// always return `Ok` and the [`ZeroRttAccepted`] will always resolve to true.
     ///
-    /// For incoming connections, a 0.5-RTT connection will always be successfully constructed.
+    /// If manually providing a [`crypto::ServerConfig`][crate::crypto::ServerConfig], check your
+    /// implementation's docs for 0-RTT pitfalls.
+    ///
+    /// ## Security
+    ///
+    /// On outgoing connections, this enables transmission of 0-RTT data, which is vulnerable to
+    /// replay attacks, and should therefore never invoke non-idempotent operations.
+    ///
+    /// On incoming connections, this enables transmission of 0.5-RTT data, which is vulnerable
+    /// to an attack in which a man-in-the-middle may prevent the handshake from successfully
+    /// completing.
```

Comment:
I don't understand what distinct attack this is describing. A MITM can always fail a connection at any time by dropping traffic.

Reply:
Oh, I was just deferring to the pre-existing comment added by @djc in [this commit](https://github.com/quinn-rs/quinn/commit/2fb15b9a0721a46ea2336b9f079ffdd9f3154a85)

>    /// On incoming connections, this enables transmission of 0.5-RTT data, which might be
>    /// intercepted by a man-in-the-middle. If this occurs, the handshake will not complete
>    /// successfully.

It's possible I misinterpreted it. Do you know what that comment is referring to?

Reply:
My reading of that comment is that a MITM can *intercept and decode* the data, which is more serious.

Reply:
Huh, I'm not aware of that being the case. @djc I don't suppose you'd be able to recall what that comment was referring to?

Reply:
Not really -- I guess this is ultimately [`ServerConfig::send_half_rtt_data`](https://docs.rs/rustls/latest/rustls/server/struct.ServerConfig.html#structfield.send_half_rtt_data), but I don't find its documentation all that clear either:

> Whether the server should send “0.5RTT” data. This means the server sends data after its first flight of handshake messages, without waiting for the client to complete the handshake.

> This can improve TTFB latency for either server-speaks-first protocols, or client-speaks-first protocols when paired with “0RTT” data. This comes at the cost of a subtle weakening of the normal handshake integrity guarantees that TLS provides. Note that the initial ClientHello is indirectly authenticated because it is included in the transcript used to derive the keys used to encrypt the data.
> 
> This only applies to TLS1.3 connections. TLS1.2 connections cannot do this optimisation and this setting is ignored for them. It is also ignored for TLS1.3 connections that even attempt client authentication.
> 
> This defaults to false. This means the first application data sent by the server comes after receiving and validating the client’s handshake up to the `Finished` message. This is the safest option.

Reply:
See also https://bblanche.gitlabpages.inria.fr/publications/BhargavanBlanchetKobeissiSP2017.pdf (search for "0.5-RTT Weak Authentication").

Reply:
Yes a MITM can always prevent a handshake from completing (like dropping, reordering, corrupting any message in it). The 0.5RTT flag is enabling the server to send data before it can confirm the handshake was successful in the server -> client direction; that means it hasn't completed any client authentication (rustls outlaws this case) but also that any corruption in the server -> client handshake won't be detected by the server yet. That is not a problem if the data that would be sent is always 100% public (ie, an attacker could just connect to the server and get the same data).

Reply:
It sounds like the only thing we need to warn here is that client authentication can't have been performed yet.

Reply:
Thanks for the links. Helped me understand a bit more.